zerver/lib/markdown/__init__.py

```diff
@@ -1358,12 +1359,16 @@ def run(self, root: Element) -> None:
                 continue
 
             if self.zmd.url_embed_data is None or url not in self.zmd.url_embed_data:
-                self.zmd.zulip_rendering_result.links_for_preview.add(url)
-                continue
+                # See if we can pull it from the cache
+                extracted_data = cached_link_embed_data(url)
+                if extracted_data is None:
+                    self.zmd.zulip_rendering_result.links_for_preview.add(url)
+                    continue
 
             # Existing but being None means that we did process the
             # URL, but it was not valid to preview.
-            extracted_data = self.zmd.url_embed_data[url]
+            else:
+                extracted_data = self.zmd.url_embed_data[url]
```

Comment:
Does this add a `memcached` query from inside the markdown processor? I'm not sure whether that's safe.